File: Assignment_B4.cpp
#include<bits/stdc++.h>
using namespace std;

class PageReplacement{
    vector<int> pageNumbers;
    int maxNumFrames;

public:
    PageReplacement(const vector<int> &pgNos, const int &frames){
        this -> pageNumbers = pgNos;
        this -> maxNumFrames = frames;
    }

    float firstInFirstOut(){
        deque<int> dq;
        int pageFaultCount = 0;
        for(int i=0;i<pageNumbers.size();i++){
            for(int page : dq){
                cout<<page<<" ";
            }
            cout<<endl;

            if(find(dq.begin(), dq.end(), pageNumbers[i]) == dq.end()){
                pageFaultCount++;
                if(dq.size() == maxNumFrames) dq.pop_front();
                dq.push_back(pageNumbers[i]);
            }
        }
        return ((float)pageFaultCount) / (float)pageNumbers.size();
    }

    float leastRecentlyUsed(){
        vector<int> v, ages;
        int pageFaultCount = 0;

        for(int i=0;i<pageNumbers.size();i++){
            for(int page : v) cout<<page<<" ";
            cout<<endl;

            for_each(ages.begin(), ages.end(), [](int &element){
                element++;
            });

            if(find(v.begin(), v.end(), pageNumbers[i]) == v.end()){
                pageFaultCount++;
                if(v.size() == maxNumFrames){
                    int maxElementIndex = max_element(ages.begin(), ages.end()) - ages.begin();
                    ages.erase(ages.begin() + maxElementIndex);
                    v.erase(v.begin() + maxElementIndex);
                    v.insert(v.begin() + maxElementIndex, pageNumbers[i]);
                    ages.insert(ages.begin() + maxElementIndex, 0);
                }
                else{
                    v.insert(v.begin(), pageNumbers[i]);
                    ages.insert(ages.begin(), 0);
                }
            }
            else{
                int index = find(v.begin(), v.end(), pageNumbers[i]) - v.begin();
                ages[index] = 0;
            }
        }
        return ((float)pageFaultCount) / pageNumbers.size();
    }

    float optimalApproach(){
        vector<int> v;
        int pageFaultCount = 0;

        for(int i=0;i<pageNumbers.size();i++){
            for(int page : v) cout<<page<<" ";
            cout<<endl;

            if(find(v.begin(), v.end(), pageNumbers[i]) == v.end()){
                pageFaultCount++;

                if(v.size() == maxNumFrames){
                    int maxUsageGap = -1;
                    int maxUsageGapIndex = 0;

                    for(int p=0;p<maxNumFrames;p++){
                        bool isFound = false;
                        
                        for(int j = i-1;j<pageNumbers.size();j++){
                            if(pageNumbers[j] == v[p]){
                                isFound = true;
                                if((j-i) > maxUsageGap){
                                    maxUsageGap = j - i;
                                    maxUsageGapIndex = p;
                                }
                            }
                        }
                        if(!isFound){
                            maxUsageGapIndex = p;
                            break;
                        }
                    }
                    v[maxUsageGapIndex] = pageNumbers[i];
                }
                else{
                    v.push_back(pageNumbers[i]);
                }
            }
        }
        return ((float)pageFaultCount) / (float)pageNumbers.size();
    }
};

int main(){
    vector<int> frameNumbers = { 1 , 3 , 0 , 3 , 5 , 6 , 3 , 4 , 7 , 1 , 0 , 8 } ; 
    PageReplacement algo( frameNumbers , 3 ) ; 

    cout << "First In First Out (FIFO): " << '\n' ; 
    float fifoRatio = algo.firstInFirstOut() ; 
    cout << "Miss Ratio: " << fifoRatio << '\n' ; 

    cout << "Least Recently Used (LRU): " << '\n' ; 
    float lruRatio = algo.leastRecentlyUsed() ; 
    cout << "Miss Ratio: " << lruRatio << '\n' ;  

    cout << "Optimal: " << '\n' ; 
    float optimalRatio = algo.optimalApproach() ; 
    cout << "Miss Ratio: " << optimalRatio << '\n' ; 

    return 0;
}

File: Assignment_B4.py
class PageReplacementAlgorithms:
    def __init__(self, page_numbers, max_num_frames) -> None:
        self.page_numbers = page_numbers
        self.max_num_frames = max_num_frames

    def first_in_first_out(self):
        frames = []
        page_fault_count = 0

        for i in range(len(self.page_numbers)):
            print(frames, end=' '.ljust(10))

            if self.page_numbers[i] not in frames:
                print(str(self.page_numbers[i]).ljust(10)," Miss ".ljust(10))
                page_fault_count += 1
                if len(frames) == self.max_num_frames:
                    frames.pop(0)
                frames.append(self.page_numbers[i])
            else:
                print(str(self.page_numbers[i]).ljust(10)," Hit ".ljust(10))
            
        return page_fault_count / len(self.page_numbers)
    
    def least_recently_used(self):
        frames , frame_ages = [], []
        page_fault_count = 0

        for i in range(len(self.page_numbers)):
            print(frames, end=' '.ljust(10))  

            for j in range(len(frame_ages)):
                frame_ages[j] += 1

            if self.page_numbers[i] not in frames:
                print(str(self.page_numbers[i]).ljust(10)," Miss ".ljust(10))
                page_fault_count += 1
                if len(frames) == self.max_num_frames:
                    max_index = frame_ages.index(max(frame_ages))
                    frame_ages.pop(max_index)
                    frames.pop(max_index)
                    frames.insert(max_index, self.page_numbers[i])
                    frame_ages.insert(max_index, 0)
                else:
                    frames.insert(0, self.page_numbers[i])
                    frame_ages.insert(0, 0)
            else:
                print(str(self.page_numbers[i]).ljust(10)," Hit ".ljust(10))
                index = frames.index(self.page_numbers[i])
                frame_ages[index] = 0

        return page_fault_count / len(self.page_numbers)
    
    def optimal_approach(self):
        frames = []
        page_fault_count = 0

        for i in range(len(self.page_numbers)):
            print(frames, end=' '.ljust(10))

            if self.page_numbers[i] not in frames:
                print(str(self.page_numbers[i]).ljust(10)," Miss ".ljust(10))
                page_fault_count += 1

                if len(frames) == self.max_num_frames:
                    max_usage_gap = -1
                    max_usage_gap_index = 0

                    for p in range(self.max_num_frames):
                        is_found = False

                        for j in range(i-1, len(self.page_numbers)):
                            if self.page_numbers[j] == frames[p]:
                                is_found = True
                                if (j-i) > max_usage_gap:
                                    max_usage_gap = j-i
                                    max_usage_gap_index = p
                        
                        if is_found == False:
                            max_usage_gap_index = p
                            break

                    frames[max_usage_gap_index] = self.page_numbers[i]
                else:
                    frames.append(self.page_numbers[i])
            else:
                print(str(self.page_numbers[i]).ljust(10)," Hit ".ljust(10))
                
        return page_fault_count / len(self.page_numbers)
    


frames = [7,0,1,2,0,3,0,4,2,3,0,3,2,3]
algo = PageReplacementAlgorithms(frames, 4)
miss_ratio = 0

print("FIFO.....")
miss_ratio = algo.first_in_first_out()
print("Miss ratio: ",  miss_ratio)
print()

print("LRU.......")
miss_ratio = algo.least_recently_used()
print("Miss ratio: ",  miss_ratio)
print()

print("Optimal......")
miss_ratio = algo.optimal_approach()
print("Miss ratio: ",  miss_ratio)
print()
